### DISCUSSION TITLE
[A-L] (23 ottobre 2023) Esercizio 2

### DISCUSSION THREAD

**Discussion by Frisbon** [sapienzastudentsnetwork/progettazione-di-algoritmi#80] [Esami A-L]
![image](https://github.com/sapienzastudentsnetwork/progettazione-di-algoritmi/assets/84731341/a8b6bc02-8302-4fc5-96c7-3cd8a06aae96)

**Comment by cornflexxx** [comment]
### Soluzione con programmazione dinamica:
Costruiamo la matrice $T_{n\times A}$ con $T[i,j]\rightarrow$ costo minimo per ottenere valore $v_j$ con $x_i$ nel sottoinsieme.

$$T[i,j] = \left\\{ \begin{array}{cl}
c_i & v_i>j\\
\displaystyle min_{\substack{q\neq i}} \\{ T[q,j-v_i]\\})+ c_i& altrimenti
\end{array} \right.$$

```
ValoreMax_CostoMinimo(A, C, V, X)
 T : Matrice n x (A+1) inizializzata a infinito
 for i = 0,...,n-1:
     for j = 0,...,A:
         if V[i] > j:
             T[i,j] = C[i]
         else:
             m = ∞
             for q = 0,...,n-1:
                 if q != i and T[q,j-V[i]] != ∞:
                     m = min(m, T[q,j-V[i]])
             if m != ∞:
                 T[i,j] = m + C[i]

 min = ∞
 for i = 0,...,n-1:
     if T[i,A] < min:
         min = T[i,A]
 sol = {}
 sol_c, sol_v = 0, 0
 
 for i = 0,...,n-1:
     if T[i,A] == min and sol_c + C[i] <= min and sol_v <= A:
	sol.add(X[i])
 return sol
```
Complessita : $O(n^2A)$

**Comment by Elia-Belli** [comment]
## Programmazione Dinamica (simile a Knapsack's Problem)
$$T[k,a] = \text{minimo costo considerando i primi k oggetti con valore almeno a}$$
- $T[0,a] = c_0$   se $v_0 \geq a$

$$T[k,a] = \begin{cases}min(c_k, T[k-1,a])\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ v_k \geq a\\
min(T[k-1,a],c_k + T[k-1,a-v_k])\ \ \ \ \ \ \ \ \ \ altrimenti\end{cases}$$

```python
def perro_salchicha(C,V,A):
   #calcolo tabella
    T = [[inf]*(A+1)]*n #tabella n x (A+1)
    for a in range(A+1):
        if(V[0] >= a): T[0,a] = C[0]

    for k in range(1,n):
    	for a in range(1,A+1):
		if(V[k] >= a):
			T[k,a] = min(C[k],T[k-1,a])
		else:
			T[k.a] = min(T[k-1,a], C[k] + T[k-1,a-V[k]])

    #calcolo sottoinsieme
    Sol = {}
    a = A
    for k in range(n,0):
	if(T[k,a] < T[k-1,a]):
		Sol.add(k)
		a -= V[k]
			
    return Sol
```
